### PR TITLE
Integrate ESLint into our development and CI process

### PR DESCRIPTION
**Pull request by philippwaller**
This PR integrates ESLint into our development and CI process. The following changes have been made:

- Update ESLint and all dependencies 
- Integration of ESLint into the webpack build process
- Adding ESLint rules to improve code quality and readability

I have made a preselection of rules that I believe would increase the overall code quality. Of course, this selection is very opinionated, so I'm looking forward to your feedback. The following rules are already activated and conform to the current code base. 

| Rule                                          | Level | Description                                                                           |
|-----------------------------------------------|-------|---------------------------------------------------------------------------------------|
| constructor-super                             | error | [Link](https://eslint.org/docs/rules/constructor-super)                               |
| cypress/no-assigning-return-values            | error | [Link](https://github.com/cypress-io/eslint-plugin-cypress)                           |
| cypress/no-async-tests                        | error | [Link](https://github.com/cypress-io/eslint-plugin-cypress)                           |
| cypress/no-unnecessary-waiting                | error | [Link](https://github.com/cypress-io/eslint-plugin-cypress)                           |
| for-direction                                 | error | [Link](https://eslint.org/docs/rules/for-direction)                                   |
| getter-return                                 | error | [Link](https://eslint.org/docs/rules/getter-return)                                   |
| import/default                                | error | [Link](https://github.com/benmosher/eslint-plugin-import)                             |
| import/export                                 | error | [Link](https://github.com/benmosher/eslint-plugin-import)                             |
| import/named                                  | error | [Link](https://github.com/benmosher/eslint-plugin-import)                             |
| import/namespace                              | error | [Link](https://github.com/benmosher/eslint-plugin-import)                             |
| jsx-quotes                                    | error | [Link](https://eslint.org/docs/rules/jsx-quotes)                                      |
| linebreak-style                               | error | [Link](https://eslint.org/docs/rules/linebreak-style)                                 |
| no-async-promise-executor                     | error | [Link](https://eslint.org/docs/rules/no-async-promise-executor)                       |
| no-class-assign                               | error | [Link](https://eslint.org/docs/rules/no-class-assign)                                 |
| no-compare-neg-zero                           | error | [Link](https://eslint.org/docs/rules/no-compare-neg-zero)                             |
| no-cond-assign                                | error | [Link](https://eslint.org/docs/rules/no-cond-assign)                                  |
| no-const-assign                               | error | [Link](https://eslint.org/docs/rules/no-const-assign)                                 |
| no-constant-condition                         | error | [Link](https://eslint.org/docs/rules/no-constant-condition)                           |
| no-control-regex                              | error | [Link](https://eslint.org/docs/rules/no-control-regex)                                |
| no-delete-var                                 | error | [Link](https://eslint.org/docs/rules/no-delete-var)                                   |
| no-dupe-args                                  | error | [Link](https://eslint.org/docs/rules/no-dupe-args)                                    |
| no-dupe-class-members                         | error | [Link](https://eslint.org/docs/rules/no-dupe-class-members)                           |
| no-dupe-else-if                               | error | [Link](https://eslint.org/docs/rules/no-dupe-else-if)                                 |
| no-dupe-keys                                  | error | [Link](https://eslint.org/docs/rules/no-dupe-keys)                                    |
| no-duplicate-case                             | error | [Link](https://eslint.org/docs/rules/no-duplicate-case)                               |
| no-empty-character-class                      | error | [Link](https://eslint.org/docs/rules/no-empty-character-class)                        |
| no-empty-pattern                              | error | [Link](https://eslint.org/docs/rules/no-empty-pattern)                                |
| no-ex-assign                                  | error | [Link](https://eslint.org/docs/rules/no-ex-assign)                                    |
| no-extra-boolean-cast                         | error | [Link](https://eslint.org/docs/rules/no-extra-boolean-cast)                           |
| no-extra-semi                                 | error | [Link](https://eslint.org/docs/rules/no-extra-semi)                                   |
| no-fallthrough                                | error | [Link](https://eslint.org/docs/rules/no-fallthrough)                                  |
| no-func-assign                                | error | [Link](https://eslint.org/docs/rules/no-func-assign)                                  |
| no-global-assign                              | error | [Link](https://eslint.org/docs/rules/no-global-assign)                                |
| no-import-assign                              | error | [Link](https://eslint.org/docs/rules/no-import-assign)                                |
| no-inner-declarations                         | error | [Link](https://eslint.org/docs/rules/no-inner-declarations)                           |
| no-invalid-regexp                             | error | [Link](https://eslint.org/docs/rules/no-invalid-regexp)                               |
| no-irregular-whitespace                       | error | [Link](https://eslint.org/docs/rules/no-irregular-whitespace)                         |
| no-misleading-character-class                 | error | [Link](https://eslint.org/docs/rules/no-misleading-character-class)                   |
| no-mixed-spaces-and-tabs                      | error | [Link](https://eslint.org/docs/rules/no-mixed-spaces-and-tabs)                        |
| no-new-symbol                                 | error | [Link](https://eslint.org/docs/rules/no-new-symbol)                                   |
| no-obj-calls                                  | error | [Link](https://eslint.org/docs/rules/no-obj-calls)                                    |
| no-octal                                      | error | [Link](https://eslint.org/docs/rules/no-octal)                                        |
| no-prototype-builtins                         | error | [Link](https://eslint.org/docs/rules/no-prototype-builtins)                           |
| no-redeclare                                  | error | [Link](https://eslint.org/docs/rules/no-redeclare)                                    |
| no-regex-spaces                               | error | [Link](https://eslint.org/docs/rules/no-regex-spaces)                                 |
| no-self-assign                                | error | [Link](https://eslint.org/docs/rules/no-self-assign)                                  |
| no-setter-return                              | error | [Link](https://eslint.org/docs/rules/no-setter-return)                                |
| no-shadow-restricted-names                    | error | [Link](https://eslint.org/docs/rules/no-shadow-restricted-names)                      |
| no-sparse-arrays                              | error | [Link](https://eslint.org/docs/rules/no-sparse-arrays)                                |
| no-this-before-super                          | error | [Link](https://eslint.org/docs/rules/no-this-before-super)                            |
| no-undef                                      | error | [Link](https://eslint.org/docs/rules/no-undef)                                        |
| no-unexpected-multiline                       | error | [Link](https://eslint.org/docs/rules/no-unexpected-multiline)                         |
| no-unreachable                                | error | [Link](https://eslint.org/docs/rules/no-unreachable)                                  |
| no-unsafe-finally                             | error | [Link](https://eslint.org/docs/rules/no-unsafe-finally)                               |
| no-unsafe-negation                            | error | [Link](https://eslint.org/docs/rules/no-unsafe-negation)                              |
| no-unsafe-optional-chaining                   | error | [Link](https://eslint.org/docs/rules/no-unsafe-optional-chaining)                     |
| no-unused-labels                              | error | [Link](https://eslint.org/docs/rules/no-unused-labels)                                |
| no-useless-escape                             | error | [Link](https://eslint.org/docs/rules/no-useless-escape)                               |
| no-whitespace-before-property                 | error | [Link](https://eslint.org/docs/rules/no-whitespace-before-property)                   |
| no-with                                       | error | [Link](https://eslint.org/docs/rules/no-with)                                         |
| require-yield                                 | error | [Link](https://eslint.org/docs/rules/require-yield)                                   |
| space-in-parens                               | error | [Link](https://eslint.org/docs/rules/space-in-parens)                                 |
| use-isnan                                     | error | [Link](https://eslint.org/docs/rules/use-isnan)                                       |
| valid-typeof                                  | error | [Link](https://eslint.org/docs/rules/valid-typeof)                                    |
| vue/comment-directive                         | error | [Link](https://eslint.vuejs.org/rules/comment-directive.html)                         |
| vue/experimental-script-setup-vars            | error | [Link](https://eslint.vuejs.org/rules/experimental-script-setup-vars.html)            |
| vue/html-end-tags                             | warn  | [Link](https://eslint.vuejs.org/rules/html-end-tags.html)                             |
| vue/jsx-uses-vars                             | error | [Link](https://eslint.vuejs.org/rules/jsx-uses-vars.html)                             |
| vue/no-arrow-functions-in-watch               | error | [Link](https://eslint.vuejs.org/rules/no-arrow-functions-in-watch.html)               |
| vue/no-async-in-computed-properties           | error | [Link](https://eslint.vuejs.org/rules/no-async-in-computed-properties.html)           |
| vue/no-custom-modifiers-on-v-model            | error | [Link](https://eslint.vuejs.org/rules/no-custom-modifiers-on-v-model.html)            |
| vue/no-dupe-keys                              | error | [Link](https://eslint.vuejs.org/rules/no-dupe-keys.html)                              |
| vue/no-dupe-v-else-if                         | error | [Link](https://eslint.vuejs.org/rules/no-dupe-v-else-if.html)                         |
| vue/no-duplicate-attributes                   | error | [Link](https://eslint.vuejs.org/rules/no-duplicate-attributes.html)                   |
| vue/no-lone-template                          | warn  | [Link](https://eslint.vuejs.org/rules/no-lone-template.html)                          |
| vue/no-multiple-slot-args                     | warn  | [Link](https://eslint.vuejs.org/rules/no-multiple-slot-args.html)                     |
| vue/no-multiple-template-root                 | error | [Link](https://eslint.vuejs.org/rules/no-multiple-template-root.html)                 |
| vue/no-reserved-keys                          | error | [Link](https://eslint.vuejs.org/rules/no-reserved-keys.html)                          |
| vue/no-shared-component-data                  | error | [Link](https://eslint.vuejs.org/rules/no-shared-component-data.html)                  |
| vue/no-side-effects-in-computed-properties    | error | [Link](https://eslint.vuejs.org/rules/no-side-effects-in-computed-properties.html)    |
| vue/no-spaces-around-equal-signs-in-attribute | warn  | [Link](https://eslint.vuejs.org/rules/no-spaces-around-equal-signs-in-attribute.html) |
| vue/no-template-key                           | error | [Link](https://eslint.vuejs.org/rules/no-template-key.html)                           |
| vue/no-textarea-mustache                      | error | [Link](https://eslint.vuejs.org/rules/no-textarea-mustache.html)                      |
| vue/no-unused-components                      | error | [Link](https://eslint.vuejs.org/rules/no-unused-components.html)                      |
| vue/no-unused-vars                            | error | [Link](https://eslint.vuejs.org/rules/no-unused-vars.html)                            |
| vue/no-use-v-if-with-v-for                    | error | [Link](https://eslint.vuejs.org/rules/no-use-v-if-with-v-for.html)                    |
| vue/no-v-for-template-key                     | error | [Link](https://eslint.vuejs.org/rules/no-v-for-template-key.html)                     |
| vue/no-v-model-argument                       | error | [Link](https://eslint.vuejs.org/rules/no-v-model-argument.html)                       |
| vue/one-component-per-file                    | warn  | [Link](https://eslint.vuejs.org/rules/one-component-per-file.html)                    |
| vue/prop-name-casing                          | warn  | [Link](https://eslint.vuejs.org/rules/prop-name-casing.html)                          |
| vue/require-component-is                      | error | [Link](https://eslint.vuejs.org/rules/require-component-is.html)                      |
| vue/require-prop-type-constructor             | error | [Link](https://eslint.vuejs.org/rules/require-prop-type-constructor.html)             |
| vue/require-render-return                     | error | [Link](https://eslint.vuejs.org/rules/require-render-return.html)                     |
| vue/require-v-for-key                         | error | [Link](https://eslint.vuejs.org/rules/require-v-for-key.html)                         |
| vue/require-valid-default-prop                | error | [Link](https://eslint.vuejs.org/rules/require-valid-default-prop.html)                |
| vue/return-in-computed-property               | error | [Link](https://eslint.vuejs.org/rules/return-in-computed-property.html)               |
| vue/use-v-on-exact                            | error | [Link](https://eslint.vuejs.org/rules/use-v-on-exact.html)                            |
| vue/v-bind-style                              | warn  | [Link](https://eslint.vuejs.org/rules/v-bind-style.html)                              |
| vue/valid-template-root                       | error | [Link](https://eslint.vuejs.org/rules/valid-template-root.html)                       |
| vue/valid-v-bind                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-bind.html)                              |
| vue/valid-v-bind-sync                         | error | [Link](https://eslint.vuejs.org/rules/valid-v-bind-sync.html)                         |
| vue/valid-v-cloak                             | error | [Link](https://eslint.vuejs.org/rules/valid-v-cloak.html)                             |
| vue/valid-v-else                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-else.html)                              |
| vue/valid-v-else-if                           | error | [Link](https://eslint.vuejs.org/rules/valid-v-else-if.html)                           |
| vue/valid-v-for                               | error | [Link](https://eslint.vuejs.org/rules/valid-v-for.html)                               |
| vue/valid-v-html                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-html.html)                              |
| vue/valid-v-if                                | error | [Link](https://eslint.vuejs.org/rules/valid-v-if.html)                                |
| vue/valid-v-model                             | error | [Link](https://eslint.vuejs.org/rules/valid-v-model.html)                             |
| vue/valid-v-on                                | error | [Link](https://eslint.vuejs.org/rules/valid-v-on.html)                                |
| vue/valid-v-once                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-once.html)                              |
| vue/valid-v-pre                               | error | [Link](https://eslint.vuejs.org/rules/valid-v-pre.html)                               |
| vue/valid-v-show                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-show.html)                              |
| vue/valid-v-text                              | error | [Link](https://eslint.vuejs.org/rules/valid-v-text.html)                              |
| vue/no-v-html                               | off | [Link](https://eslint.vuejs.org/rules/no-v-html.html)                               |

The following rules would currently fail and are therefore temporarily disabled. However, I am of the opinion that these should also be activated in the medium term. Due to the large number of required changes, the activations should be clustered into multiple pull requests. I would follow up on that after we have a common agreement.

| Rule                                        | Level         | Description                                                                         |
|---------------------------------------------|---------------|-------------------------------------------------------------------------------------|
| camelcase                                   | error         | [Link](https://eslint.org/docs/rules/camelcase)                                     |
| comma-dangle                                | error         | [Link](https://eslint.org/docs/rules/comma-dangle)                                  |
| comma-spacing                               | error         | [Link](https://eslint.org/docs/rules/comma-spacing)                                 |
| eol-last                                    | error         | [Link](https://eslint.org/docs/rules/eol-last)                                      |
| indent                                      | error, 2      | [Link](https://eslint.org/docs/rules/indent)                                        |
| no-console                                    | error         | [Link](https://eslint.org/docs/rules/no-console)                                      |
| no-debugger                                    | error         | [Link](https://eslint.org/docs/rules/no-debugger)                                      |
| no-empty                                    | error         | [Link](https://eslint.org/docs/rules/no-empty)                                      |
| no-trailing-spaces                          | error         | [Link](https://eslint.org/docs/rules/no-trailing-spaces)                            |
| no-unused-vars                              | error         | [Link](https://eslint.org/docs/rules/no-unused-vars)                                |
| no-useless-catch                            | error         | [Link](https://eslint.org/docs/rules/no-useless-catch)                              |
| quotes                                      | error, single | [Link](https://eslint.org/docs/rules/quotes)                                        |
| vue/attributes-order                        | error         | [Link](https://eslint.vuejs.org/rules/attributes-order.html)                        |
| vue/component-tags-order                    | error         | [Link](https://eslint.vuejs.org/rules/component-tags-order.html)                    |
| vue/html-closing-bracket-newline            | error         | [Link](https://eslint.vuejs.org/rules/html-closing-bracket-newline.html)            |
| vue/html-closing-bracket-spacing            | error         | [Link](https://eslint.vuejs.org/rules/html-closing-bracket-spacing.html)            |
| vue/html-indent                             | error         | [Link](https://eslint.vuejs.org/rules/html-indent.html)                             |
| vue/html-quotes                             | error         | [Link](https://eslint.vuejs.org/rules/html-quotes.html)                             |
| vue/html-self-closing                       | error         | [Link](https://eslint.vuejs.org/rules/html-self-closing.html)                       |
| vue/max-attributes-per-line                 | error         | [Link](https://eslint.vuejs.org/rules/max-attributes-per-line.html)                 |
| vue/multiline-html-element-content-newline  | error         | [Link](https://eslint.vuejs.org/rules/multiline-html-element-content-newline.html)  |
| vue/mustache-interpolation-spacing          | error         | [Link](https://eslint.vuejs.org/rules/mustache-interpolation-spacing.html)          |
| vue/no-multi-spaces                         | error         | [Link](https://eslint.vuejs.org/rules/no-multi-spaces.html)                         |
| vue/no-mutating-props                       | error         | [Link](https://eslint.vuejs.org/rules/no-mutating-props.html)                       |
| vue/no-parsing-error                        | error         | [Link](https://eslint.vuejs.org/rules/no-parsing-error.html)                        |
| vue/no-template-shadow                      | error         | [Link](https://eslint.vuejs.org/rules/no-template-shadow.html)                      |
| vue/order-in-components                     | error         | [Link](https://eslint.vuejs.org/rules/order-in-components.html)                     |
| vue/require-default-prop                    | error         | [Link](https://eslint.vuejs.org/rules/require-default-prop.html)                    |
| vue/require-prop-types                      | error         | [Link](https://eslint.vuejs.org/rules/require-prop-types.html)                      |
| vue/singleline-html-element-content-newline | error         | [Link](https://eslint.vuejs.org/rules/singleline-html-element-content-newline.html) |
| vue/this-in-template                        | error         | [Link](https://eslint.vuejs.org/rules/this-in-template.html)                        |
| vue/v-on-style                              | error         | [Link](https://eslint.vuejs.org/rules/v-on-style.html)                              |
| vue/v-slot-style                            | error         | [Link](https://eslint.vuejs.org/rules/v-slot-style.html)                            |
| vue/valid-v-slot                            | error         | [Link](https://eslint.vuejs.org/rules/valid-v-slot.html)                            |

**Attention:** After merging, these changes can impact the build success of pending pull requests.

### CODE REVIEW

**Comment by ghys** [comment]
Thanks for taking care of this!
I was content with just having ESLint hints in the IDE with the extension, but since I had to remind PR authors a couple times to verify there were no linting errors, I suppose it's time to have it more strictly enforced.

About the second table of rules, some of them do indeed make sense but I'm not hesitant about those:
- no-console
- those related to attributes: order, max per line etc.
- the default component tags order (I prefer "template","style","script")
- enforcing types and defaults in props
- no-v-html cannot be enabled, we require it to display config parameter descriptions

**Comment by philippwaller** [comment]
Here are my two cents:

**no console**
Loggin to the console might be great during development but should be avoided in production. It will reduce the overall application performance (especially in loops), increase the memory consumption and may reveal information that is not intended for the user.

**Attribute related rules like order, max per line etc**
I am a big fan of these rules. From my point of view, they ensure a consistent convention across all files and increase the overall readability. We have some Vue file whose lines have well over 200 characters. The record holder even has 376 characters per line (thing-details.vue:20). It's not fun to edit these files, even on my 32-inch 4k monitor.

**Component tags order**
Fair enough. I adjusted the rule accordingly. 

**Enforcing types and defaults in props**
Strong typing and default values can prevent bugs, specially when multiple developers involved. Take the property `thingId` as an example. As a new developer, I need to review the code before I know whether a string or an integer is expected. I know it's general JavaScript problem, but it's a step in the right direction. (again, personal opinion)

**no-v-html**
Got you. Can't be enabled.

Since we are mainly talking about the rule that will not be activated for the time being, I suggest that we merge the commit and evaluate the changes in the respective pull requests. It is difficult to assess the impact without looking at concrete examples. What do think @ghys?

**Comment by ghys** [comment]
Yes of course I surely understand the rationale behind these rules, I was mainly in "RAD" mode for the entirety of this project so relatively lax on the code tidiness and it can be a little dirty at places, these rules would make it better. I'm using the Vue devtools extension, as you might imagine, to figure out what is going on, what are props set to, etc.
As you pointed out there are most involved rules that would require a lot of changes if enabled - and the inevitable regression testing, very important - if you're willing to come up with these, I will certainly not veto any.
Personal preference, I'd still like not to completely forbid `console.log` for the occasional important message in production - most of them were replaced with `console.debug` anyways - so we can probably check there's no unnecessary logging in code reviews instead of ESLint.
For the time being, we can start with the rules that are already enabled - you'll have to rebase though. Thanks for your effort!

**Comment by philippwaller** [comment]
> we can probably check there's no unnecessary logging in code reviews instead of ESLint.
Sounds like a good alternative. I have adjusted the rules accordingly.

**Comment by ghys** [comment]
@philippwaller This seems to work well, one last little request: could you disable `linebreak-style` for the time being - this one caused me some problems. I'd prefer keeping the git behavior on Windows as it is described in https://eslint.org/docs/rules/linebreak-style#using-this-rule-with-version-control-systems which is fine, than adding a `.gitattributes` file for now.
We can consider re-adding it later.
Otherwise, LGTM!

**Review by ghys — APPROVED** [review]
Many thanks!

File: bundles/org.openhab.ui/web/.eslintrc.js [hunk 1]
```diff
@@ -2,23 +2,27 @@ module.exports = {
...
+    browser: true,
+    node: true,
+    es6: true
   },
 
   extends: [
-    // https://github.com/vuejs/eslint-plugin-vue#priority-a-essential-error-prevention
-    // consider switching to `plugin:vue/strongly-recommended` or `plugin:vue/recommended` for stricter rules.
-    'plugin:vue/essential',
-    '@vue/standard'
```
Comment (ghys) [inline bundles/org.openhab.ui/web/.eslintrc.js:17]: @philippwaller I just noticed you removed @vue/standard here!? Any reason why?

Comment (ghys) [reply bundles/org.openhab.ui/web/.eslintrc.js:17]: Hm, it appears it's much more strict than before...

**Pull request merged**